content/posts/controller-exception-handling.md

```diff
@@ -0,0 +1,93 @@
+This behaviour can be changed in the `apps/[your_app_name]/application.rb` file, just look for `handle_exception` in the environment specific configuration blocks.
+
+The default behaviour is fine if you are rendering HTML pages, however, for APIs you should always return responses in the respective format, e.g., JSON or XML.
+Let's assume you are implementing a JSON API for managing a Library. Your current task is to implement a checkout procedure and you need to decide what happens when two patrons try to borrow the last available copy of a book.
+Thanks to ACID guarantees of your database you are sure that when you book the last copy for one patron, the other will fail to book it. However, you need to gracefully handle this scenario.
+
+Your controller could look something like this:
+
+```ruby
+# apps/inventory_api/controllers/loans/create.rb
+module InventoryApi
+  module Controllers
+    module Loans
+      class Create
+        include Web::Action
+
+        handle_exception Inventory::OutOfStock => :borrowing_failed
+        expose :loan
+        
+        params do
+          required(:book_id).filled
+          required(:patron_id).filled
+        end
+
+        def call(params)
+          patron = PatronRepository.new.find(params[:patron_id])
+          book = BookRepository.new.find(params[:book_id])
+          # calling borrow may raise Inventory::OutOfStock
+          @loan = Inventory.borrow(book, patron)
+        end
+
+        def borrowing_failed(_exception)
+          error = {
+            status: "not_found",
+            code: "out_of_stock",
+            title: "No more copies of the requested book are available",
+          }
+
+          halt 404, JSON.dump(errors: [error])
+        end
+      end
+    end
+  end
+end
+```
+
+By calling `handle_exception Inventory::OutOfStock => :borrowing_failed` we tell Hanami to call `borrowing_failed` when the `Inventory::OutOfStock` exception is raised. The `borrowing_failed` method is then handed the exception object for further inspection, but we don't actually need that object in our example.
+Instead we simply render a [JSON:API](https://jsonapi.org/format/) error response to tell the client that no copy is left for borrowing.
+
+Since we wish to document that this actually works, we write a test for this behaviour. However, when we run our test the expected JSON response is not rendered, instead we are presented with a stack trace.
+
+We remeber that this is due to Hanami's default behaviour in face of exceptions. For our test to succeed we need to change how Hanami handles exceptions in the test environment for this controller.
+We could change it for all controllers of this application with the following configuration:
+
+```ruby
+# apps/inventory_api/application.rb
+
+module InventoryApi
```

Comment:
Can we put this code snippet higher in article so it will be picked as the cover? It's more explicative than that action. Thanks.

Reply:
I agree, but I'll have to think about restructuring the article accordingly.